client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java

```diff
@@ -0,0 +1,508 @@
+                    return Float.parseFloat(parser.getText());
+                default:
+                    return illegalToken(parser);
+            }
+        }
+    }
+
+    private static class BooleanDecoder
+            implements TypeDecoder
+    {
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.nextToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_FALSE:
+                    return false;
+                case VALUE_TRUE:
+                    return true;
+                case VALUE_STRING:
+                    return Boolean.parseBoolean(parser.getText());
+                default:
+                    return illegalToken(parser);
+            }
+        }
+    }
+
+    private static class StringDecoder
+            implements TypeDecoder
+    {
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.nextToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_STRING:
+                    return parser.getText();
+                default:
+                    return illegalToken(parser);
+            }
+        }
+    }
+
+    private static class Base64Decoder
+            implements TypeDecoder
+    {
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.nextToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_STRING:
+                    return Base64.getDecoder().decode(parser.getText());
+                default:
+                    return illegalToken(parser);
+            }
+        }
+    }
+
+    private static class ArrayDecoder
+            implements TypeDecoder
+    {
+        private final TypeDecoder typeDecoder;
+
+        public ArrayDecoder(ClientTypeSignature signature)
+        {
+            requireNonNull(signature, "signature is null");
+            checkArgument(signature.getRawType().equals(ARRAY), "not an array type signature: %s", signature);
+            this.typeDecoder = createTypeDecoder(signature.getArgumentsAsTypeSignatures().get(0));
+        }
+
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.nextToken()) {
+                case START_ARRAY:
+                    break;
+                case VALUE_NULL:
+                    return null;
+                default:
+                    return illegalToken(parser);
+            }
+            List<Object> values = new ArrayList<>(); // nulls allowed
+            while (true) {
+                try {
+                    values.add(typeDecoder.decode(parser));
+                }
+                catch (IllegalTokenException e) {
```

Comment:
Using exceptions for control flow here seems bad. Also, it's expensive, because we construct the exception with a formatted message and stack trace.

What if we change the flow so that `TypeDecoder` expects a current token in the `JsonParser`? Then `ArrayDecoder` would know to stop when `nextToken() == END_ARRAY` and wouldn't call the child decoder?

Reply:
Yeah, that makes more sense. I've switched to the assumption that type decoder always moves to the current token to be decoded.